code/src/components/Todos.js

```diff
@@ -0,0 +1,91 @@
+import React from 'react';
+import styled from 'styled-components';
+import { useSelector, useDispatch } from 'react-redux';
+import { Todo } from './Todo';
+import { visibilityFilter } from 'reducers/visibilityFilter';
+
+const StyledTodos = styled.div`
+  display: ${props => (props.empty ? 'flex' : 'block')}
+  flex-direction: ${props => (props.empty ? 'column' : '')}
+  justify-content: ${props => (props.empty ? 'center' : '')};
+  align-items: ${props => (props.empty ? 'center' : '')};
+  height: ${props => (props.empty ? '200px' : '')};
+  padding: 0 0.4rem;
+`;
+
+const Message = styled.p`
+  display: block;
+  font-size: 1.3rem;
+  text-align: center;
+  font-style: italic;
+`;
+
+const Icon = styled.div`
+  display: flex;
+  justify-content: center;
+  align-items: center;
+  margin: 2rem 0 1.5rem 0;
+  font-size: 4rem;
+  color: rgba(230, 0, 60, 1);
+`;
+
+export const Todos = () => {
+  const dispatch = useDispatch();
+  const filter = useSelector(state => state.visibilityFilter);
+  let allTodos = useSelector(state => state.todos);
+  let filteredTodos = null;
+
+  switch (filter) {
+    case 'SHOW_ACTIVE':
+      filteredTodos = allTodos
+        .filter(todo => todo.completed === false)
+        .sort((a, b) => b.pinned - a.pinned);
+      if (filteredTodos.length === 0) {
+        dispatch(
+          visibilityFilter.actions.setVisibility({ filter: 'SHOW_ALL' })
+        );
+      }
+      break;
+    case 'SHOW_COMPLETED':
+      filteredTodos = allTodos
+        .filter(todo => todo.completed === true)
+        .sort((a, b) => b.pinned - a.pinned);
+      if (filteredTodos.length === 0) {
+        dispatch(
+          visibilityFilter.actions.setVisibility({ filter: 'SHOW_ALL' })
+        );
+      }
+      break;
+    case 'SHOW_PINNED':
+      filteredTodos = allTodos.filter(todo => todo.pinned === true);
+      if (filteredTodos.length === 0) {
+        dispatch(
+          visibilityFilter.actions.setVisibility({ filter: 'SHOW_ALL' })
+        );
+      }
+      break;
+    default:
+      filteredTodos = allTodos
+        .filter(todo => todo)
+        .sort((a, b) => b.pinned - a.pinned);
+  }
```

Comment:
There's a whole lot going on here. You should aim to keep your components as dumb as possible. The side effects (changing filter if there's nothing to show) should happen in a reducer, and the all the filtering could be moved into fancier selectors, or potentially a custom hook. Something like this, perhaps:

```js
// In another file
const useFilteredTodosSelector = () => {
  const filter = useSelector(state => state.visibilityFilter)

  switch (filter) {
    case 'SHOW_ACTIVE':
      return useSelector(state => state.todos.filter(...).sort(...))
    case 'SHOW_COMPLETED':
      return useSelector(state => state.todos.filter(...).sort(...))
    ...
  }
}

// then, in this component:

const todos = useFilteredTodosSelector()
```

Reply:
Looking at the code again I agree there is a lot going on. But I always want to improve myself so I will research and come up with a better solution. 

Despite that I'm happy I managed to implement the logic I had for my filter idea.